sam_view.c

```diff
@@ -560,211 +560,412 @@ static const char *copied_tags[] = { "RG", "BC", "QT", NULL };
+                || (next->core.flag&(state->flag_off)) != 0) continue;
+            ++n_reads;
+
+            if (!strcmp(bam_get_qname(current), bam_get_qname(next))) { // Is a pair
+                // Translate read into fastq format into the output buffer
+                if (bam1_to_fq(current, &linebuf, state) == false) return true;
+                fputs(linebuf.s, state->fpr[which_readpart(current)]);
+                // Translate read into fastq format into the output buffer
+                if (bam1_to_fq(next, &linebuf, state) == false) return true;
+                fputs(linebuf.s, state->fpr[which_readpart(next)]);
+                break;
+            } else { // current read is singleton put next read into current and reread.
+                ++n_singletons;
+                // Translate read into fastq format into the output buffer
+                if (bam1_to_fq(current, &linebuf, state) == false) return true;
+                fputs(linebuf.s, state->fpse);
+                // buffer swap
+                bam1_t* tmp = current;
+                current = next;
+                next = tmp;
             }
-            kputs((char*)buf, &linebuf);
-            kputc('\n', &linebuf);
         }
     }
+    free(linebuf.s);
+    bam_destroy1(current);
+    bam_destroy1(next);
+    
+    fprintf(stderr, "[M::%s] discarded %" PRId64 " singletons\n", __func__, n_singletons);
+    fprintf(stderr, "[M::%s] processed %" PRId64 " reads\n", __func__, n_reads);
+    return false;
+}
 
-    if (fpse) {
-        if ( previous ) { // Nothing left to match it's a singleton
-            ++n_singletons;
-            fputs(linebuf.s, fpse);  // Write previous read to singletons
-        } else {
-            fputs(linebuf.s, stdout); // Write previous read
-        }
+static bool bam2fq_mainloop(bam2fq_state_t *state)
+{
+    // process a name collated BAM into fastq
+    bam1_t* b = bam_init1();
+    if (b == NULL) {
+        perror(NULL);
+        return true;
+    }
+    int64_t n_reads = 0; // Statistics
+    kstring_t linebuf = { 0, 0, NULL }; // Buffer
+    while (sam_read1(state->fp, state->h, b) >= 0) {
+        if (b->core.flag&(BAM_FSECONDARY|BAM_FSUPPLEMENTARY) // skip secondary and supplementary alignments
+            || (b->core.flag&(state->flag_on)) != state->flag_on             // or reads indicated by filter flags
+            || (b->core.flag&(state->flag_off)) != 0) continue;
+        ++n_reads;
 
-        fprintf(stderr, "[M::%s] discarded %" PRId64 " singletons\n", __func__, n_singletons);
-        fclose(fpse);
-    } else {
-        fputs(linebuf.s, stdout); // Write previous read
+        /////////////////////////////////////////////////////////////
+        // Translate read into fastq format into the output buffer //
+        /////////////////////////////////////////////////////////////
+        if (bam1_to_fq(b, &linebuf, state) == false) return true;
```

Comment:
Some functions return `false` to indicate failure, other functions return `true` to indicate failure.  It would be good to sort this out.  Also prefer `if (! foo())` to `if (foo() == false)`.